### PR TITLE
Add support for bulk marking of attendance

### LINKED ISSUE #150: Add support for bulk marking of attendance
Attendance-marking is restricted to a single index or tutorial group.

Let's
* Add support for bulk marking of attendance, using a similar notation as bulk delete

### PR DESCRIPTION
**Pull request by philbertshea**
Attendance-marking is restricted to a single index or tutorial group.

Let's

* Add support for bulk marking of attendance, using a similar notation as bulk delete

Closes #150.

### CODE REVIEW

File: docs/DeveloperGuide.md [hunk 1]
```diff
@@ -420,7 +420,7 @@ For all use cases below, the **System** is the `TAssist` and the **Actor** is th
   Use case resumes at step 2.
 
 * 1c. User provides at least one argument that is invalid, or in incorrect format.
-  For instance, user provides an index that is out of range.
+  For instance, user provides an indexList that is out of range.
```
Comment (ChinZJ) [inline docs/DeveloperGuide.md:423]: Not sure if indexList is considered implementation details, if so better to simply call it index. What do you think?

File: src/main/java/seedu/tassist/logic/parser/ParserUtil.java [hunk 2]
```diff
@@ -299,6 +301,36 @@ public static List<Index> parseMultipleIndexes(String input) throws ParseExcepti
...
 
+    /**
+     * Parses an input string representing a list of TutGroups ("T01,T02", "T02-T05")
+     * and returns a sorted, de-duplicated list of {@link TutGroup} objects.
+     *
+     * @param input Input string provided.
+     * @return List of TutGroups parsed from the list.
+     * @throws ParseException If the input is invalid.
+     */
+    public static List<TutGroup> parseMultipleTutGroups(String input) throws ParseException {
```
Comment (ChinZJ) [inline src/main/java/seedu/tassist/logic/parser/ParserUtil.java:312]: Unsure how feasible, but possible for all bulk ops to share a common method?

ue the method will take in a prefix then it will perform its checks.
Delete command will have no prefix, tut will have T

**Review by ChinZJ — COMMENTED** [review]
Minor nitpick on implementation details for consideration.

File: docs/UserGuide.md [hunk 3]
```diff
@@ -180,8 +180,8 @@ Deletes the specified person from the address book.
 Format: `del -i INDEX`
 
 * Deletes the person at the specified `INDEX`.
-* The index refers to the index number shown in the displayed person list.
-* The index **must be a positive integer** 1, 2, 3, …​
+* The indexList refers to the indexList number shown in the displayed person list.
```
Comment (hhlh27) [inline docs/UserGuide.md:183]: I think should put indexes instead of indexList in the user guide, since indexList is an implementation detail

File: src/test/java/seedu/tassist/logic/commands/MarkAttendanceCommandTest.java [hunk 4]
```diff
@@ -365,23 +367,24 @@ public void equals() {
...
-                        VALID_WEEK_B, Attendance.NO_TUTORIAL);
+                new MarkAttendanceCommand(VALID_WEEK_B, Attendance.NO_TUTORIAL,
+                        List.of(new TutGroup(VALID_TUT_GROUP_AMY)));
 
         // Different tut group -> returns false.
-        assertFalse(standardCommandWithTutGroup.equals(new MarkAttendanceCommand(
-                new TutGroup(VALID_TUT_GROUP_BOB), VALID_WEEK_B, Attendance.NO_TUTORIAL)));
+        assertFalse(standardCommandWithTutGroup.equals(
+                new MarkAttendanceCommand(VALID_WEEK_B, Attendance.NO_TUTORIAL,
+                List.of(new TutGroup(VALID_TUT_GROUP_BOB)))));
```
Comment (hhlh27) [inline src/test/java/seedu/tassist/logic/commands/MarkAttendanceCommandTest.java:387]: Good handling of edge cases

File: docs/DeveloperGuide.md (hunk 1 above)
Comment (philbertshea) [reply docs/DeveloperGuide.md:423]: Good catch, will change this back to index

File: src/main/java/seedu/tassist/logic/parser/ParserUtil.java (hunk 2 above)
Comment (philbertshea) [reply src/main/java/seedu/tassist/logic/parser/ParserUtil.java:312]: Hmm. I think I am sharing Hannah's parseMultipleIndexes method, for multiple indexes -i 1,2,3
The difference is that I am parsing multiple tutorial groups here too, because remember the mark attendance command also supports marking for the whole tutorial group. for multiple tutgroups -t T01,T02,T03

Because it's called "parseMultipleIndexes", I think it makes sense to have another method to parse tutorial groups rather than indexes. The implementations call other methods to handle tutGroups too

File: src/test/java/seedu/tassist/logic/commands/MarkAttendanceCommandTest.java [hunk 5]
```diff
@@ -239,7 +241,7 @@ public void execute_markTutGroupNotAttendedUnfilteredList_success() {
         }
 
         MarkAttendanceCommand command =
-                new MarkAttendanceCommand(tutGroupToEdit, replacementIndex + 1, Attendance.NOT_ATTENDED);
+                new MarkAttendanceCommand(replacementIndex + 1, Attendance.NOT_ATTENDED, List.of(tutGroupToEdit));
```
Comment (aquaimpact) [inline src/test/java/seedu/tassist/logic/commands/MarkAttendanceCommandTest.java:244]: Interesting way to prevent clashing!

**Review by aquaimpact — APPROVED** [review]
Overall, LGTM!

File: src/main/java/seedu/tassist/logic/commands/MarkAttendanceCommand.java [hunk 6]
```diff
@@ -109,31 +109,31 @@ public class MarkAttendanceCommand extends Command {
...
      *
-     * @param tutGroup Tutorial group to mark attendance for. Should be null if index is not null.
+     * @param tutGroupList List of Tutorial groups to mark attendance for. Should be null if index is not null.
      * @param week Week to mark attendance of person for.
      * @param attendanceStatus New Attendance Status to set the person or tutorial group to.
      */
-    public MarkAttendanceCommand(TutGroup tutGroup, int week, int attendanceStatus) {
-        requireAllNonNull(tutGroup, week, attendanceStatus);
-        this.index = null;
+    public MarkAttendanceCommand(int week, int attendanceStatus, List<TutGroup> tutGroupList) {
```
[the remark this answers is not in the document]
Comment (philbertshea) [reply src/main/java/seedu/tassist/logic/commands/MarkAttendanceCommand.java:132]: It's a way to circumvent type erasure. 
I tried alternatives like varargs, but it would require packing and unpacking.

Could probably try arrays in the future at v1.5.
